parameter/criterion/src/Criterion.cpp

```diff
@@ -0,0 +1,296 @@
+    } else {
+        // Name
+        strFormattedDescription = "Criterion name: " + mName;
+
+        if (bWithTypeInfo) {
+            // Type Kind
+            strFormattedDescription += ", type kind: ";
+            strFormattedDescription +=  typeName;
+        }
+
+        // Current State
+        strFormattedDescription += ", current state: " + getFormattedState();
+
+        if (bWithTypeInfo) {
+            // States
+            strFormattedDescription += ", states: " + listPossibleValues();
+        }
+    }
+    return strFormattedDescription;
+}
+
+void Criterion::toXml(CXmlElement& xmlElement, CXmlSerializingContext& serializingContext) const
+{
+    (void)serializingContext;
+    xmlElement.setAttributeString("Value", getFormattedState());
+    xmlElement.setAttributeString("Name", mName);
+    xmlElement.setAttributeString("Kind", isInclusive() ? "Inclusive" : "Exclusive");
+
+    for (auto& valuePair : mValues) {
+        CXmlElement childValuePairElement;
+
+        xmlElement.createChild(childValuePairElement, "ValuePair");
+        childValuePairElement.setAttributeString("Literal", valuePair.first);
+        childValuePairElement.setAttributeSignedInteger("Numerical", valuePair.second);
+    }
+}
+
+bool Criterion::isInclusive() const
+{
+    return false;
+}
+
+bool Criterion::getNumericalValue(const std::string& literalValue,
+                                  int& numericalValue) const
+{
+    try {
+        numericalValue = mValues.at(literalValue);
+        return true;
+    }
+    catch (std::out_of_range& e) {
+        return false;
+    }
+}
+
+bool Criterion::getLiteralValue(int numericalValue, std::string& literalValue) const
+{
+    for (auto& value : mValues) {
+        if (value.second == numericalValue) {
+            literalValue = value.first;
+            return true;
+        }
+    }
+    return false;
+}
+
+std::string Criterion::getFormattedState() const
+{
+    assert(!mState.empty());
+    std::string formattedState;
+    bool succeed = getLiteralValue(*(mState.begin()), formattedState);
+    assert(succeed);
+    return formattedState;
+}
+
+std::string Criterion::listPossibleValues() const
+{
+    std::string possibleValues = "{";
+
+    // Get comma separated list of values
+    bool first = true;
+    for (auto& value : mValues) {
+
+        if (first) {
+            first = false;
+        } else {
+            possibleValues += ", ";
+        }
+        possibleValues += value.first;
+    }
+    possibleValues += "}";
+
+    return possibleValues;
+}
+
+bool Criterion::match(const std::string& method, const State& state) const
+{
+    return mMatchMethods.at(method)(state);
+}
+
+bool Criterion::isMatchMethodAvailable(const std::string& method) const
+{
+    return mMatchMethods.count(method) == 1;
+}
+
+template<class MapType>
+MapType Criterion::gatherMap(MapType mapA, const MapType& mapB)
+{
+    mapA.insert(mapB.begin(), mapB.end());
+    return mapA;
+}
```

Comment:
non-virtual protected "addMatchMethod" method: seems ok to me but we loose the initialization list use (but does it really matter?)

Suggestion 2: Exclusive criterion match method would be duplicated no ?